crates/astria-composer/src/searcher/executor.rs

```diff
@@ -120,178 +139,266 @@ impl Executor {
     ///
     /// # Errors
     /// An error is returned if connecting to the sequencer fails.
     pub(super) async fn run_until_stopped(mut self) -> eyre::Result<()> {
-        // set up connection to sequencer
-        self.init_nonce_from_sequencer(5, Duration::from_secs(5), 2.0)
+        let mut submission_fut = Fuse::terminated();
+        self.nonce = get_latest_nonce(self.sequencer_client.clone(), self.address)
             .await
-            .wrap_err("failed retrieving initial nonce from sequencer")?;
+            .wrap_err("failed getting initial nonce from sequencer")?;
+        self.status.send_modify(|status| status.is_connected = true);
+        loop {
+            select! {
+                biased;
 
-        while let Some(bundle) = self.executor_rx.recv().await {
-            if let Err(e) = self.sign_and_submit(bundle).await {
-                // FIXME: currently this will fail both when there is an issue with the nonce and
-                // when unable to reach the sequencer. As there is currently no error returned by
-                // the sequencer for invalid nonces, there is nothing to handle. This should be
-                // changed after #364 is merged in a followup PR to handle nonce failues.
-                error!(
-                    error.message = %e,
-                    error.cause_chain = ?e,
-                    "failed submitting the bundle to the sequencer; bailing",
-                );
-                break;
+                rsp = &mut submission_fut, if !submission_fut.is_terminated() => {
+                    match rsp {
+                        Ok(new_nonce) => self.nonce = new_nonce,
+                        Err(_e) => bail!("eyeyey"),
+                    }
+                }
+
+                // receive new bundle for processing
+                Some(bundle) = self.executor_rx.recv(), if submission_fut.is_terminated() => {
+                    // TODO(https://github.com/astriaorg/astria/issues/476): Attach the hash of the
+                    // bundle to the span. Linked GH issue is for agreeing on a hash for `SignedTransaction`,
+                    // but both should be addressed.
+                    let span =  info_span!(
+                        "execute bundle",
+                        address = %self.address,
+                        nonce.initial = self.nonce,
+                        bundle.len = bundle.len(),
+                    );
+                    submission_fut = SubmitFut {
+                        client: self.sequencer_client.clone(),
+                        address: self.address,
+                        nonce: self.nonce,
+                        signing_key: self.sequencer_key.clone(),
+                        state: SubmitState::NotStarted,
+                        bundle,
+                    }
+                    .instrument(span)
+                    .fuse();
+                }
```

Comment:
doesn't this overwrite `submission_fut` is we get successive bundles from `executor_rx` before the previous `submission_fut` is done?

Reply:
No, because the executor_ex is conditional on the submit fut not being terminated (hence the reason why we use fuses).

Reply:
missed that haha